File: android/app/src/main/java/blog/nextlab/echo/ui/chatlist/NewGroupScreen.kt
package blog.nextlab.echo.ui.chatlist

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.runtime.toMutableStateList
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.platform.LocalSoftwareKeyboardController
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.unit.dp
import blog.nextlab.echo.ui.common.ScreenHeader
import blog.nextlab.echo.core.designsystem.RinowaDimens
import blog.nextlab.echo.core.designsystem.RinowaTheme
import blog.nextlab.echo.core.designsystem.glassFace
import blog.nextlab.echo.core.haptics.HapticToken
import blog.nextlab.echo.core.haptics.LocalRinowaHaptics
import blog.nextlab.echo.model.ConversationId
import blog.nextlab.echo.model.UserId
import blog.nextlab.echo.ui.auth.RinowaField
import blog.nextlab.echo.ui.auth.NoticeBanner
import blog.nextlab.echo.ui.auth.PrimaryButton
import blog.nextlab.echo.ui.auth.PrimaryButtonLabel
import blog.nextlab.echo.ui.common.Avatar

/**
 * グループを作る。
 *
 * 候補に出るのは、このアカウントがすでに会話したことのある人
 * （[ChatListViewModel.contacts]）。先に追加する友達一覧は別に無い。Rinowa では、
 * 話したことがあることが知っていること。一覧に居ない人へは招待コードで1回届けば、
 * 次からは居る。
 */
@Composable
fun NewGroupScreen(
    viewModel: ChatListViewModel,
    onCreated: (ConversationId) -> Unit,
    onBack: () -> Unit,
) {
    val colors = RinowaTheme.colors
    val type = RinowaTheme.type
    val haptics = LocalRinowaHaptics.current
    val keyboard = LocalSoftwareKeyboardController.current

    var title by remember { mutableStateOf("") }
    var busy by remember { mutableStateOf(false) }
    var notice by remember { mutableStateOf<String?>(null) }
    val selected = remember { mutableListOf<UserId>().toMutableStateList() }

    val contacts = viewModel.contacts

    fun create() {
        if (busy) return
        keyboard?.hide()
        busy = true
        notice = null
        viewModel.createGroup(title.trim(), selected.toList()) { outcome ->
            busy = false
            when (outcome) {
                is StartChatOutcome.Opened -> {
                    haptics.perform(HapticToken.Success)
                    onCreated(outcome.id)
                }

                else -> {
                    haptics.perform(HapticToken.Error)
                    notice = "作成できませんでした。通信を確認してください。"
                }
            }
        }
    }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(colors.background)
            .statusBarsPadding()
            .imePadding(),
    ) {
        ScreenHeader(title = "グループを作る", onBack = {
            keyboard?.hide()
            onBack()
        })

        Column(Modifier.padding(horizontal = 24.dp)) {
            RinowaField(
                value = title,
                onValueChange = { title = it.take(60) },
                placeholder = "グループ名（例: 家族）",
                enabled = !busy,
                keyboardOptions = KeyboardOptions(imeAction = ImeAction.Done),
                keyboardActions = KeyboardActions(),
            )
            Spacer(Modifier.height(16.dp))
            Text(
                text = if (selected.isEmpty()) {
                    "追加する人を選んでください"
                } else {
                    "${selected.size}人を追加します"
                },
                style = type.label,
                color = colors.textSecondary,
            )
            Spacer(Modifier.height(8.dp))
        }

        if (contacts.isEmpty()) {
            Column(
                modifier = Modifier
                    .weight(1f)
                    .fillMaxWidth()
                    .padding(horizontal = 40.dp),
                verticalArrangement = Arrangement.Center,
                horizontalAlignment = Alignment.CenterHorizontally,
            ) {
                Text(
                    text = "まだ誰とも話していません",
                    style = type.listName,
                    color = colors.textPrimary,
                )
                Spacer(Modifier.height(8.dp))
                Text(
                    text = "先に招待コードで1対1の会話をはじめると、その相手をグループに入れられます。",
                    style = type.listPreview,
                    color = colors.textSecondary,
                    textAlign = androidx.compose.ui.text.style.TextAlign.Center,
                )
            }
        } else {
            LazyColumn(Modifier.weight(1f).fillMaxWidth()) {
                items(contacts, key = { it.id.value }) { profile ->
                    val isSelected = profile.id in selected
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(
                                horizontal = RinowaDimens.glassCardMargin,
                                vertical = RinowaDimens.glassCardGap / 2,
                            )
                            .glassFace(shape = RoundedCornerShape(RinowaDimens.glassCorner))
                            .clickable {
                                haptics.perform(HapticToken.Selection)
                                if (isSelected) selected.remove(profile.id)
                                else selected.add(profile.id)
                            }
                            .padding(RinowaDimens.listItemPadding),
                    ) {
                        Avatar(title = profile.displayName, seed = profile.avatarSeed)
                        Spacer(Modifier.width(12.dp))
                        Text(
                            text = profile.displayName,
                            style = type.listName,
                            color = colors.textPrimary,
                            modifier = Modifier.weight(1f),
                        )
                        SelectionMark(selected = isSelected)
                    }
                }
            }
        }

        Column(Modifier.padding(horizontal = 24.dp)) {
            if (notice != null) Spacer(Modifier.height(12.dp))
            NoticeBanner(text = notice, isError = true)
            Spacer(Modifier.height(12.dp))
            PrimaryButton(
                enabled = !busy && selected.isNotEmpty() && title.isNotBlank(),
                onClick = ::create,
            ) { color ->
                PrimaryButtonLabel(if (busy) "作成しています" else "グループを作る", color)
            }
            Spacer(Modifier.height(20.dp))
            Spacer(Modifier.navigationBarsPadding())
        }
    }
}

@Composable
private fun SelectionMark(selected: Boolean) {
    val colors = RinowaTheme.colors
    Box(
        modifier = Modifier
            .size(24.dp)
            .clip(CircleShape)
            .background(if (selected) colors.accent else colors.surfaceSunken),
        contentAlignment = Alignment.Center,
    ) {
        if (!selected) return@Box
        Canvas(Modifier.size(14.dp)) {
            val stroke = Stroke(width = 2.dp.toPx(), cap = StrokeCap.Round, join = StrokeJoin.Round)
            val w = size.width
            val h = size.height
            val path = Path().apply {
                moveTo(w * 0.16f, h * 0.54f)
                lineTo(w * 0.42f, h * 0.78f)
                lineTo(w * 0.84f, h * 0.24f)
            }
            drawPath(path, colors.onAccent, style = stroke)
        }
    }
}

File: ios/App/Sources/Screens/NewGroupScreen.swift
import SwiftUI

/// グループを作る。
///
/// `ui/chatlist/NewGroupScreen.kt` の Swift 側。文言はそのまま写した。
///
/// 候補に出るのは、このアカウントがすでに会話したことのある人。
/// **先に追加する友達一覧は別に無い。** Rinowa では、話したことがあることが
/// 知っていること。一覧に居ない人へは招待コードで1回届けば、次からは居る。
struct NewGroupScreen: View {

    let onBack: () -> Void
    var onCreated: (String) -> Void = { _ in }

    @EnvironmentObject private var store: ConversationStore
    @Environment(\.rinowaColors) private var colors
    @Environment(\.haptics) private var haptics

    @State private var title = ""
    @State private var busy = false
    @State private var notice: String?
    @State private var selected: Set<String> = []

    private var contacts: [Contact] { store.contacts }

    var body: some View {
        VStack(spacing: 0) {
            ScreenHeader(title: "グループを作る", onBack: onBack)

            VStack(alignment: .leading, spacing: 0) {
                RinowaField(
                    value: Binding(
                        get: { title },
                        set: { title = String($0.prefix(60)) }
                    ),
                    placeholder: "グループ名（例: 家族）",
                    enabled: !busy
                )
                Spacer().frame(height: 16)
                Text(selected.isEmpty ? "追加する人を選んでください" : "\(selected.count)人を追加します")
                    .rinowaType(RinowaType.label)
                    .foregroundStyle(colors.textSecondary)
                Spacer().frame(height: 8)
            }
            .padding(.horizontal, 24)

            if contacts.isEmpty {
                empty
            } else {
                list
            }

            VStack(spacing: 0) {
                if notice != nil { Spacer().frame(height: 12) }
                NoticeBanner(text: notice)
                Spacer().frame(height: 12)
                PrimaryButton(
                    enabled: !busy && !selected.isEmpty
                        && !title.trimmingCharacters(in: .whitespaces).isEmpty,
                    action: create
                ) { tint in
                    PrimaryButtonLabel(text: busy ? "作成しています" : "グループを作る", color: tint)
                }
                Spacer().frame(height: 20)
            }
            .padding(.horizontal, 24)
        }
        .background(colors.background.ignoresSafeArea())
        .navigationBarBackButtonHidden()
        .toolbar(.hidden, for: .navigationBar)
    }

    private var empty: some View {
        VStack(spacing: 8) {
            Text("まだ誰とも話していません")
                .rinowaType(RinowaType.listName)
                .foregroundStyle(colors.textPrimary)
            Text("先に招待コードで1対1の会話をはじめると、その相手をグループに入れられます。")
                .rinowaType(RinowaType.listPreview)
                .foregroundStyle(colors.textSecondary)
                .multilineTextAlignment(.center)
        }
        .padding(.horizontal, 40)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }

    private var list: some View {
        ScrollView {
            LazyVStack(spacing: RinowaDimens.glassCardGap) {
                ForEach(contacts) { contact in
                    let isSelected = selected.contains(contact.id)
                    HStack(spacing: 12) {
                        Avatar(title: contact.displayName, seed: contact.seed)
                        Text(contact.displayName)
                            .rinowaType(RinowaType.listName)
                            .foregroundStyle(colors.textPrimary)
                            .frame(maxWidth: .infinity, alignment: .leading)
                        SelectionMark(selected: isSelected)
                    }
                    .padding(RinowaDimens.rowPadding)
                    .frame(maxWidth: .infinity)
                    .glassFace()
                    .padding(.horizontal, RinowaDimens.glassCardMargin)
                    .contentShape(Rectangle())
                    .onTapGesture {
                        haptics.fire(.selection)
                        if isSelected { selected.remove(contact.id) }
                        else { selected.insert(contact.id) }
                    }
                }
            }
            .padding(.vertical, RinowaDimens.glassCardGap)
        }
        .frame(maxHeight: .infinity)
    }

    private func create() {
        guard !busy else { return }
        busy = true
        notice = nil
        // 作る仕組みは iOS にまだ無い。失敗と同じ形で返す。
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.4) {
            busy = false
            haptics.fire(.error)
            notice = "作成できませんでした。通信を確認してください。"
        }
    }
}

/// 選んだ印。
///
/// 塗りつぶした丸に、白い鉤。**枠だけのチェックボックスにしない。**
/// 遠目に「入っているか」が読めるのは、線ではなく面。
private struct SelectionMark: View {
    let selected: Bool
    @Environment(\.rinowaColors) private var colors

    var body: some View {
        ZStack {
            Circle().fill(selected ? colors.accent : colors.surfaceSunken)
            if selected {
                Canvas { context, size in
                    let w = size.width, h = size.height
                    var path = Path()
                    path.move(to: CGPoint(x: w * 0.16, y: h * 0.54))
                    path.addLine(to: CGPoint(x: w * 0.42, y: h * 0.78))
                    path.addLine(to: CGPoint(x: w * 0.84, y: h * 0.24))
                    context.stroke(
                        path,
                        with: .color(colors.onAccent),
                        style: StrokeStyle(lineWidth: 2, lineCap: .round, lineJoin: .round)
                    )
                }
                .frame(width: 14, height: 14)
            }
        }
        .frame(width: 24, height: 24)
        .animation(RinowaMotion.pop, value: selected)
    }
}
